File: spark/bruteforce.cpp
/*************************************************************************
	> Author: Wayne Ho
	> Purpose: brute-force counting triangles
	> Created Time: Fri Jul 17 01:08:33 2015
	> Mail: hewr2010@gmail.com 
 ************************************************************************/
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;

vector<pair<int, int> > edges;
map<pair<int, int>, bool> pool;
int n, m;

int main(int argc, char **argv) {
	scanf("%d%d", &n, &m);
	for (int i = 0; i < m; ++i) {
		int x, y;
		scanf("%d%d", &x, &y);
		auto key = make_pair(x, y);
		pool[key] = true;
		edges.push_back(key);
	}
	sort(edges.begin(), edges.end());
    // count triangles
    int cnt(0);
    for (int start = 0; start < edges.size(); ) {
        int end = start;
        while (end < edges.size() && edges[end].first == edges[start].first) ++end;
        for (int i = start; i != end; ++i) for (int j = i + 1; j != end; ++j) {
            auto key = make_pair(edges[i].second, edges[j].second);
            if (pool.find(key) != pool.end()) ++cnt;
        }
        start = end;
    }
	printf("%d\n", cnt);
    return 0;
}

File: spark/bruteforce.py
#!/usr/bin/python
# -*- coding:utf-8 -*-
# Created Time: 2015年07月24日 星期五 02时29分01秒
# Purpose: bruteforce
# Mail: hewr2010@gmail.com
__author__ = "Wayne Ho"

import argparse
parser = argparse.ArgumentParser()
parser.add_argument("--input", "-i",
                    help="input graph file",
                    required=False,
                    default="./input.txt",
                    type=str)
args = parser.parse_args()

if __name__ == "__main__":
    fin = open(args.input)
    fin.next()
    edges = []
    for line in fin:
        x, y = map(lambda x: int(x), line[:-1].split(" "))
        if x > y:
            x, y = y, x
        if x == y:
            continue
        edges.append((x, y))
    edges.sort()
    doc = {(x, y):True for x, y in edges}
    cnt = 0
    i = 0
    while i < len(edges):
        j = i
        while j < len(edges) and edges[j][0] == edges[i][0]:
            j += 1
        for l in xrange(i, j):
            for r in xrange(i, l):
                x, y, z = edges[l][0], edges[l][1], edges[r][1]
                if y > z:
                    y, z = z, y
                if (y, z) in doc:
                    cnt += 1
        i = j
    print cnt
